gh-138122: Add sampling profiler visualisation to docs

lkollar opened the pull request:
<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
gh-NNNNNN: Summary of the changes made
```

Where: gh-NNNNNN refers to the GitHub issue number.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `main`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNNNN)
```

Where: [X.Y] is the branch name, for example: [3.13].

GH-NNNNNN refers to the PR number from `main`.

-->

Add a visualisation explaining how the sampling profiler works.

<img width="1217" height="702" alt="image" src="https://github.com/user-attachments/assets/0c61521d-84cb-477a-a512-b8795f129265" />


CC: @pablogsal 

<!-- gh-issue-number: gh-138122 -->
* Issue: gh-138122
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--142772.org.readthedocs.build/en/142772/library/profiling.sampling.html

<!-- readthedocs-preview cpython-previews end -->

lkollar committed:
Add sampling profiler visualisation

Add an interactive JS visualisation to demonstrate how the sampling
profiler works. A small demo program is shown executing line by line,
demonstrating how stack frames are created and destroyed and how samples
are taken periodically during the process.

A new Sphinx extension is used to capture and inject the trace into the
JS file during documentation build to make these traces reproducible
(and avoid hard coding them in the source code).

StanFromIreland commented on Doc/library/profiling.sampling.rst:
```diff
@@ -44,6 +44,23 @@ of samples over a profiling session, Tachyon constructs an accurate statistical
+.. only:: html
+
+   The following interactive visualization demonstrates how sampling profiling
+   works. Press **Play** to watch a Python program execute, and observe how the
```
See the [`:guilabel:`](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-guilabel) role for buttons.

lkollar replied:
I tried using this but it just wraps the element in a `span` and has no associated styling, so the text isn't highlighted in any way. I could style it in the local CSS but this should probably get a global style rule. 

The only other place where this is used is in the [logging cookbook](https://docs.python.org/3/howto/logging-cookbook.html#running-a-logging-socket-listener-in-production) (search for "Download ZIP") which isn't styled either. Seems like a bug to me.

lkollar committed:
Fix unclosed tag

lkollar committed:
Fix epub export issue

Include the JS and CSS files in the Sphinx extension to make sure Sphinx
properly includes these when producing all output types.

pablogsal committed:
Simplify sampling profiler visualization code

Moved function colors to a lookup map and easing definitions to a module
constant so they aren't recreated on each call. Cached the background
element reference in DOMStackFrame instead of querying it on every
hover. Chained the regex replacements in syntax highlighting into one
expression.

Removed the CSS color rules for specific functions since colors are now
set in JS, along with a duplicate color variable and two methods that
were never called.
